### PR TITLE
Allow providing an external scrollview.

### PR DESCRIPTION
**Pull request by gorhom**
## Description

Fixes (issue #)

This PR will allow passing an external scrollview to the `recyclerlistview`, which is needed to allow gesture interactions with `react-native-gesture-handler` especially on **Android**.

## Reviewers’ hat-rack :tophat:

- Prop description and docs.

#### Before 

https://user-images.githubusercontent.com/4061838/177045508-f5b21abc-7604-4292-9835-c8c26018afd4.mov

#### After 

https://user-images.githubusercontent.com/4061838/177045515-b311fcbd-c06b-43b0-9254-b0ca9be93620.mov
 
## Checklist

- [x] I have added a changelog entry following the [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) guidelines.

### CODE REVIEW

**Comment by nandorojo** [comment]
Thoughts on renaming the prop to `ScrollViewComponent`?

**Comment by gorhom** [comment]
@naqvitalha sounds better! i have updated the pr 👍

File: src/FlashListProps.ts [hunk 1]
```diff
@@ -122,6 +124,14 @@ export interface FlashListProps<TItem> extends ScrollViewProps {
    */
   ListHeaderComponentStyle?: StyleProp<ViewStyle> | undefined;
 
+  /**
+   * Rendered as the main scrollview. Its contract for the scroll event should match the native scroll event contract, i.e.
+   * `scrollEvent = { nativeEvent: { contentOffset: { x: offset, y: offset } } }`
+   */
+  renderScrollComponent?:
```
Comment (naqvitalha) [inline src/FlashListProps.ts:131]: Let's use default scroll view props from RN. That should be fine. All this complexity in RLV is due to its web only implementation. FlashList doesn't depend on it so using ScrollView props should be fine.

File: src/FlashList.tsx [hunk 2]
```diff
@@ -350,6 +351,7 @@ class FlashList<T> extends React.PureComponent<
           windowCorrectionConfig={this.getUpdatedWindowCorrectionConfig()}
           itemAnimator={this.itemAnimator}
           suppressBoundedSizeException
+          externalScrollView={renderScrollComponent as any}
```
Comment (naqvitalha) [inline src/FlashList.tsx:354]: Typecast to BaseScrollView instead

File: documentation/docs/fundamentals/usage.md [hunk 3]
```diff
@@ -96,6 +96,14 @@ estimatedItemSize?: number;
 
 ---
 
+### `renderScrollComponent`
```
Comment (naqvitalha) [inline documentation/docs/fundamentals/usage.md:99]: nit: Props are sorted alphabetically. Moving this would be helpful.

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,7 @@ and adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 ## [1.0.4] - 2022-07-02
 
 - Build fix for Android projects having `kotlinVersion` defined in `build.gradle`.
+- Allow providing an external scrollview.
```
Comment (naqvitalha) [inline CHANGELOG.md:13]: Add link to this PR here

**Review by naqvitalha — COMMENTED** [review]
Few more comment. Looks good otherwise.

File: src/FlashList.tsx (hunk 2 above)
Comment (gorhom) [reply src/FlashList.tsx]: i had to cast it to `RecyclerListViewProps["externalScrollView"]` instead

```ts
    externalScrollView?: {
        new (props: ScrollViewDefaultProps): BaseScrollView;
    };
```

**Comment by gorhom** [comment]
@naqvitalha i have addressed you comments in the latest commit 👍

**Review by naqvitalha — APPROVED** [review]
Thanks for the PR. Looks great!

**Comment by sdfricke** [comment]
> Using this prop this way: `renderScrollComponent={props => <ScrollView {...props} />}` I get this warning: `Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?`

I'm seeing the same issue

**Comment by hirbod** [comment]
```tsx
const ScrollViewWithRef = React.forwardRef((props, ref) => (
  <ScrollView {...props} ref={ref} />
)) as typeof ScrollView;
```

**Comment by sdfricke** [comment]
> renderScrollComponent={props => <ScrollView {...props} />}

Thank you for the suggestion.  I'm not quite grasping how that would be implemented in the context of this simple example:
```renderScrollComponent={props => <ScrollView {...props} />}```

I've tried a couple different ways but getting quite a few errors so I'm guessing I'm not implementing that suggestion as intended.

**Comment by sdfricke** [comment]
> ```tsx
> const ScrollViewWithRef = React.forwardRef((props, ref) => (
>   <ScrollView {...props} ref={ref} />
> )) as typeof ScrollView;
> ```

Thank you for the suggestion. I'm not quite grasping how that would be implemented in the context of this simple example:
renderScrollComponent={props => <ScrollView {...props} />}

I've tried a couple different ways but getting quite a few errors so I'm guessing I'm not implementing that suggestion as intended.

**Comment by hirbod** [comment]
Please create a repo on Snack so I can have a look

**Comment by sdfricke** [comment]
@hirbod it turns out I was on older version.  It looks like the issue I was seeing was fixed in version 1.4.0 as described here: https://github.com/Shopify/flash-list/issues/616

**Pull request merged**